File: Week_01/Seminar/Solutions/Easy/PlusOne.cpp
class Solution
{
public:
    vector<int> plusOne(vector<int> &digits)
    {
        int iter = digits.size() - 1;
        int takeOver = 1;
        while (iter >= 0 && takeOver > 0)
        {
            digits[iter]++;
            digits[iter] %= 10;
            takeOver = (digits[iter] == 0);
            iter--;
        }

        if (iter == -1 && takeOver)
        {
            digits.insert(digits.begin(), 1);
        }

        return digits;
    }
};

File: Week_01/Practicum/python/Plus One.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        increment = 1
        carry = 0
        for i in range(len(digits) - 1, -1, -1):
            new_digit = (digits[i] + increment + carry) % 10
            carry = (digits[i] + increment + carry) // 10
            digits[i] = new_digit
            increment = 0 # Increment only the last digit

        if carry:
            return [1] + digits
        return digits
        
        # One-liner solution
        # return list(map(int, list(str(int("".join(map(str, digits))) + 1))))